File: frontend/objective-c/knownLanguages/knownLanguages/DataController.m
//
//  DataController.m
//  knownLanguages
//
//  Created by Tan Junhe on 27/8/23.
//

#import "DataController.h"
#import "SceneDelegate.h"

@implementation IPSetting

+ (instancetype)sharedInstance {
    static IPSetting *sharedInstance = nil;
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
            sharedInstance = [[IPSetting alloc] init];
        });
        return sharedInstance;
}   

- (NSString *)setUrl {
    IPSetting *ipsetting = [IPSetting sharedInstance];
    
    if(ipsetting.IPAddress.length == 0 && ipsetting.IPAddress.length == 0){
        NSString *configFileDir = @"config/config";//change this to base unless config.plist is created
        NSString *configFileURL = [[NSBundle mainBundle] pathForResource:configFileDir ofType:@"plist"];
        NSData *configData = [[NSFileManager defaultManager] contentsAtPath:configFileURL];
        NSError *error;
        NSPropertyListFormat plistFormat;
        NSDictionary *config = [NSPropertyListSerialization propertyListWithData:configData options:NSPropertyListImmutable format:&plistFormat error:&error];
            
        ipsetting.IPAddress = [config valueForKey:@"ipaddress"];
        ipsetting.Port = [config valueForKey:@"port"];
    }
    return [NSString stringWithFormat:@"http://%@:%@/",ipsetting.IPAddress,ipsetting.Port];
}

@end

@implementation EndSession

+ (instancetype)sharedInstance {
    static EndSession *sharedInstance = nil;
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
            sharedInstance = [[EndSession alloc] init];
        });
        return sharedInstance;
}

- (void)logout {
    NSString *bundleIdentifier = [[NSBundle mainBundle] bundleIdentifier];
    [[NSUserDefaults standardUserDefaults] removePersistentDomainForName:bundleIdentifier];
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    UIViewController *loginVC = [storyboard instantiateViewControllerWithIdentifier:@"loginVC"];
    SceneDelegate *sceneDelegate = (SceneDelegate *)[[[UIApplication sharedApplication] connectedScenes].allObjects.firstObject delegate];
    [sceneDelegate changeRootViewController:loginVC];

}

@end


@implementation UserGender

+ (instancetype)sharedInstance {
    static UserGender *sharedInstance = nil;
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
            sharedInstance = [[UserGender alloc] init];
            
        });
        return sharedInstance;
}

+ (void)initialize{
    [[UserGender sharedInstance] setGender:[[NSArray alloc] initWithObjects:@"Prefer not to say", @"Male", @"Female", @"Others", nil]];
}
@end


@implementation FormatDate

- (NSString *)formatDateSQL:(NSDate *)date {
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat:@"yyyy-MM-dd"];
    return [formatter stringFromDate:date];
}

- (NSString *)formatDate:(NSDate *)date {
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat:@"dd MMMM yyy"];
    return [formatter stringFromDate:date];
}

- (NSDate *)dateFromISOString:(NSString *)dateString {
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat:@"MM/dd/yyyy"];
    return [formatter dateFromString:dateString] ?: [NSDate date];
}

- (NSDate *)dateForPV:(NSString *)dateString {
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat:@"dd MMMM yyyy"];
    return [formatter dateFromString:dateString] ?: [NSDate date];
}

@end

File: frontend/swift/knownLanguages/knownLanguages/DataController.swift
//
//  DataController.swift
//  knownLanguages
//
//  Created by Tan Junhe on 29/7/23.
//

import Foundation
import UIKit

class IPSetting{
    static let sharedInstance = IPSetting()
    var IPAddres = ""
    var Port = ""
    var configFileDir = "config/config"//change this to base unless config.plist is created
    
    func setUrl() -> String{
        let ipSetting = IPSetting.sharedInstance
        if(ipSetting.IPAddres.count == 0 && ipSetting.Port.count == 0){
            guard let configFileURL = Bundle.main.url(forResource: configFileDir, withExtension: "plist"),
                  let configData = try? Data(contentsOf: configFileURL),
                  let config = try? PropertyListSerialization.propertyList(from: configData, format: nil) as? [String: Any]
            else {
                fatalError("Failed to load config.plist")
            }
            ipSetting.IPAddres = config["ipaddress"] as? String ?? ""
            ipSetting.Port = config["port"] as? String ?? ""
            
        }
        return "http://\(ipSetting.IPAddres):\(ipSetting.Port)/"
    }
}

class EndSession{
    static let sharedInstance = EndSession()
    
    func logout(){
        UserDefaults.standard.removePersistentDomain(forName: Bundle.main.bundleIdentifier!)
        let storyboard = UIStoryboard(name: "Main", bundle: nil)
        let loginVC = storyboard.instantiateViewController(identifier: "loginVC")
        (UIApplication.shared.connectedScenes.first?.delegate as? SceneDelegate)?.changeRootViewController(loginVC)
    }
}

class UserGender{
    static let shareInstance = UserGender()
    
    var gender = [ "Prefer not to say", "Male", "Female", "Others"]
}

class FormatDate{
    
    func formatDateSQL(date: Date) -> String{
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: date)
    }
    
    func formatDate(date: Date) -> String{
        let formatter = DateFormatter()
        formatter.dateFormat = "dd MMMM yyyy"
        return formatter.string(from: date)
    }
    
    func dateFromISOString(dateString: String) -> Date {
        let formatter = DateFormatter()
        formatter.dateFormat = "MM/dd/yyyy"
        return formatter.date(from: dateString) ?? Date()
    }

    func dateForPV(date: String)->Date{
        let formatter = DateFormatter()
        formatter.dateFormat = "dd MMMM yyyy"
        return formatter.date(from: date) ?? Date()
    }
}
